// xdl/build/detach/IosIPABuilder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createIPABuilder;

var _lodash;

function _load_lodash() {
  return _lodash = _interopRequireDefault(require('lodash'));
}

var _os = _interopRequireDefault(require('os'));

var _path = _interopRequireDefault(require('path'));

var _fsExtra;

function _load_fsExtra() {
  return _fsExtra = _interopRequireDefault(require('fs-extra'));
}

var _plist;

function _load_plist() {
  return _plist = _interopRequireDefault(require('plist'));
}

var _Logger;

function _load_Logger() {
  return _Logger = _interopRequireDefault(require('./Logger'));
}

var _ExponentTools;

function _load_ExponentTools() {
  return _ExponentTools = require('./ExponentTools');
}

var _IosCodeSigning;

function _load_IosCodeSigning() {
  return _IosCodeSigning = _interopRequireWildcard(require('./IosCodeSigning'));
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const logger = (_Logger || _load_Logger()).default.withFields({ buildPhase: 'building and signing IPA' });

function createIPABuilder(buildParams) {
  let build = (() => {
    var _ref = _asyncToGenerator(function* () {
      const { provisioningProfilePath, clientBuild } = buildParams;

      yield copyProvisioningProfileToHomedir(provisioningProfilePath, appUUID);
      logger.info('provisioning profile copied to home directory');
      const plistData = yield readCMSMessage(provisioningProfilePath);
      logger.info('done retrieving provisioning profile data');

      logger.info('checking if teamID is present in keychain and that certificate is valid...');
      const codeSignIdentity = yield (_IosCodeSigning || _load_IosCodeSigning()).ensureCertificateValid(buildParams);
      logger.info('ensured certificate is valid');

      logger.info('validating provisioning profile...');
      (_IosCodeSigning || _load_IosCodeSigning()).validateProvisioningProfile(plistData, {
        distCertFingerprint: codeSignIdentity,
        teamID,
        bundleIdentifier
      });
      logger.info('provisioning profile is valid');

      logger.info('writing export-options.plist file...');
      const exportMethod = (_IosCodeSigning || _load_IosCodeSigning()).resolveExportMethod(plistData);
      const exportOptionsPlistPath = _path.default.join(provisionDir, 'export-options.plist');
      const exportOptionsData = {
        bundleIdentifier,
        provisioningProfileUUID: plistData.UUID,
        exportMethod,
        teamID
      };
      yield (_IosCodeSigning || _load_IosCodeSigning()).writeExportOptionsPlistFile(exportOptionsPlistPath, exportOptionsData);
      logger.info('created export-options.plist file');

      logger.info('generating IPA...');
      const unsignedIpaPath = _path.default.join(buildDir, `${appUUID}-unsigned.ipa`);
      const ipaBuilderArgs = {
        ipaPath: unsignedIpaPath,
        workspace,
        archivePath: outputPath,
        codeSignIdentity,
        exportOptionsPlistPath,
        plistData,
        keychainPath,
        exportMethod
      };
      yield (_IosCodeSigning || _load_IosCodeSigning()).buildIPA(ipaBuilderArgs, buildParams, clientBuild);
      logger.info('generated IPA');

      logger.info('creating entitlements file...');
      const generatedEntitlementsPath = _path.default.join(appDir, 'generatedEntitlements.entitlements');
      yield (_IosCodeSigning || _load_IosCodeSigning()).createEntitlementsFile({
        generatedEntitlementsPath,
        plistData,
        archivePath: outputPath,
        manifest
      });
      logger.info('created entitlements file');

      logger.info('resigning IPA...');
      yield (_IosCodeSigning || _load_IosCodeSigning()).resignIPA({
        codeSignIdentity,
        entitlementsPath: generatedEntitlementsPath,
        provisioningProfilePath,
        sourceIpaPath: unsignedIpaPath,
        destIpaPath: uploadPath,
        keychainPath
      }, buildParams);
      logger.info('resigned IPA');
    });

    return function build() {
      return _ref.apply(this, arguments);
    };
  })();

  let cleanup = (() => {
    var _ref2 = _asyncToGenerator(function* () {
      try {
        yield (_fsExtra || _load_fsExtra()).default.remove(getProvisioningProfilePath(appUUID));
      } catch (err) {
        logger.error('failed to perform cleanup, error:', err);
      }
    });

    return function cleanup() {
      return _ref2.apply(this, arguments);
    };
  })();

  let copyProvisioningProfileToHomedir = (() => {
    var _ref3 = _asyncToGenerator(function* (provisioningProfilePath, appUUID) {
      yield (_fsExtra || _load_fsExtra()).default.mkdirp(getProvisioningProfileDirPath());
      const newProvisioningProfilePath = getProvisioningProfilePath(appUUID);
      yield (_fsExtra || _load_fsExtra()).default.copy(provisioningProfilePath, newProvisioningProfilePath);
    });

    return function copyProvisioningProfileToHomedir(_x, _x2) {
      return _ref3.apply(this, arguments);
    };
  })();

  let readCMSMessage = (() => {
    var _ref4 = _asyncToGenerator(function* (provisioningProfilePath) {
      const { output } = yield (0, (_ExponentTools || _load_ExponentTools()).spawnAsyncThrowError)('security', ['cms', '-D', '-i', provisioningProfilePath], {
        stdio: 'pipe'
      });
      const plistRaw = output.join('');
      const plistData = (_lodash || _load_lodash()).default.attempt((_plist || _load_plist()).default.parse, plistRaw);
      if ((_lodash || _load_lodash()).default.isError(plistData)) {
        throw new Error(`Error when parsing plist: ${plistData.message}`);
      }
      return plistData;
    });

    return function readCMSMessage(_x3) {
      return _ref4.apply(this, arguments);
    };
  })();

  const { appUUID, keychainPath, bundleIdentifier, teamID, workingDir, manifest } = buildParams;
  const workspace = _path.default.join(workingDir, 'shellAppWorkspaces', 'ios', 'default', 'ExpoKitApp.xcworkspace');
  const appDir = _path.default.join('/private/tmp/turtle', appUUID);
  const buildDir = _path.default.join(appDir, 'build');
  const provisionDir = _path.default.join(appDir, 'provisioning');
  const outputPath = _path.default.join(appDir, 'archive.xcarchive');
  const uploadPath = _path.default.join(buildDir, 'archive.ipa');

  const getProvisioningProfileDirPath = () => _path.default.join(_os.default.homedir(), 'Library/MobileDevice/Provisioning Profiles');

  const getProvisioningProfilePath = appUUID => _path.default.join(getProvisioningProfileDirPath(), `${appUUID}.mobileprovision`);

  return { build, cleanup };
}
