// xdl/build/detach/IosShellApp.js
// Copyright 2015-present 650 Industries. All rights reserved.

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createTurtleWorkspaceAsync = exports.configureAndCopyArchiveAsync = exports.buildAndCopyArtifactAsync = exports.DEFAULT_EXPOKIT_WORKSPACE_NAME = undefined;

/**
 *  Build the iOS workspace at the given path.
 *  @return the path to the resulting build artifact
 */
let _buildAsync = (() => {
  var _ref = _asyncToGenerator(function* (projectName, workspacePath, configuration, type, relativeBuildDestination, verbose) {
    let buildCmd, pathToArtifact;
    const buildDest = `${relativeBuildDestination}-${type}`;
    if (type === 'simulator') {
      buildCmd = `xcodebuild -workspace ${projectName}.xcworkspace -scheme ${projectName} -sdk iphonesimulator -configuration ${configuration} -derivedDataPath ${buildDest} CODE_SIGN_IDENTITY="" CODE_SIGNING_REQUIRED=NO ARCHS="i386 x86_64" ONLY_ACTIVE_ARCH=NO | xcpretty`;
      pathToArtifact = _path.default.join(buildDest, 'Build', 'Products', `${configuration}-iphonesimulator`, `${projectName}.app`);
    } else if (type === 'archive') {
      buildCmd = `xcodebuild -workspace ${projectName}.xcworkspace -scheme ${projectName} -sdk iphoneos -destination generic/platform=iOS -configuration ${configuration} archive -derivedDataPath ${buildDest} -archivePath ${buildDest}/${projectName}.xcarchive CODE_SIGN_IDENTITY="" CODE_SIGNING_REQUIRED=NO | xcpretty`;
      pathToArtifact = _path.default.join(buildDest, `${projectName}.xcarchive`);
    } else {
      throw new Error(`Unsupported build type: ${type}`);
    }

    (_Logger || _load_Logger()).default.info(`Building iOS workspace at ${workspacePath} to ${buildDest}:\n`);
    (_Logger || _load_Logger()).default.info(buildCmd);
    if (!verbose) {
      (_Logger || _load_Logger()).default.info('\nxcodebuild is running. Logging errors only. To see full output, use --verbose 1...');
    }
    yield (0, (_ExponentTools || _load_ExponentTools()).spawnAsyncThrowError)(buildCmd, null, {
      // only stderr
      stdio: verbose ? 'inherit' : ['ignore', 'ignore', 'inherit'],
      cwd: workspacePath,
      shell: true
    });
    return _path.default.resolve(workspacePath, pathToArtifact);
  });

  return function _buildAsync(_x, _x2, _x3, _x4, _x5, _x6) {
    return _ref.apply(this, arguments);
  };
})();

let _podInstallAsync = (() => {
  var _ref2 = _asyncToGenerator(function* (workspacePath, isRepoUpdateEnabled) {
    // ensure pods are clean
    const pathsToClean = [_path.default.join(workspacePath, 'Pods'), _path.default.join(workspacePath, 'Podfile.lock')];
    pathsToClean.forEach(function (path) {
      if ((_fsExtra || _load_fsExtra()).default.existsSync(path)) {
        (_rimraf || _load_rimraf()).default.sync(path);
      }
    });

    // install
    let cocoapodsArgs = ['install'];
    if (isRepoUpdateEnabled) {
      cocoapodsArgs.push('--repo-update');
    }
    (_Logger || _load_Logger()).default.info('Installing iOS workspace dependencies...');
    (_Logger || _load_Logger()).default.info(`pod ${cocoapodsArgs.join(' ')}`);
    yield (0, (_ExponentTools || _load_ExponentTools()).spawnAsyncThrowError)('pod', cocoapodsArgs, {
      stdio: 'inherit',
      cwd: workspacePath
    });
  });

  return function _podInstallAsync(_x7, _x8) {
    return _ref2.apply(this, arguments);
  };
})();

/**
 * @param workspacePath optionally provide a path for the unbuilt xcode workspace to create/use.
 * @param expoSourcePath path to expo client app sourcecode (/ios dir from expo/expo repo)
 * @param shellAppSdkVersion sdk version for shell app
 */


let _createStandaloneContextAsync = (() => {
  var _ref3 = _asyncToGenerator(function* (args) {
    // right now we only ever build a single detached workspace for service contexts.
    // TODO: support multiple different pod configurations, assemble a cache of those builds.
    const expoSourcePath = args.expoSourcePath || '../ios';
    let workspaceSourcePath;
    if (args.workspacePath) {
      workspaceSourcePath = args.workspacePath;
    } else {
      workspaceSourcePath = _path.default.join(expoSourcePath, '..', 'shellAppWorkspaces', 'ios', 'default');
    }
    let { privateConfigFile, privateConfigData } = args;

    let privateConfig;
    if (privateConfigData) {
      privateConfig = privateConfigData;
    } else if (privateConfigFile) {
      let privateConfigContents = yield (_fsExtra || _load_fsExtra()).default.readFile(privateConfigFile, 'utf8');
      privateConfig = JSON.parse(privateConfigContents);
    }

    let manifest;
    if (args.manifest) {
      manifest = args.manifest;
      (_Logger || _load_Logger()).default.withFields({ buildPhase: 'reading manifest' }).info('Using manifest:', JSON.stringify(manifest));
    } else if (args.url && args.sdkVersion) {
      const { url, sdkVersion, releaseChannel } = args;
      manifest = yield (0, (_ExponentTools || _load_ExponentTools()).getManifestAsync)(url, {
        'Exponent-SDK-Version': sdkVersion,
        'Exponent-Platform': 'ios',
        'Expo-Release-Channel': releaseChannel ? releaseChannel : 'default',
        Accept: 'application/expo+json,application/json'
      });
    }

    const buildFlags = (_StandaloneBuildFlags || _load_StandaloneBuildFlags()).default.createIos(args.configuration, {
      workspaceSourcePath,
      appleTeamId: args.appleTeamId
    });
    const context = (_StandaloneContext || _load_StandaloneContext()).default.createServiceContext(expoSourcePath, args.archivePath, manifest, privateConfig, args.testEnvironment, buildFlags, args.url, args.releaseChannel, args.shellAppSdkVersion);
    return context;
  });

  return function _createStandaloneContextAsync(_x9) {
    return _ref3.apply(this, arguments);
  };
})();

/**
 * possible args:
 *  @param url manifest url for shell experience
 *  @param sdkVersion sdk to use when requesting the manifest
 *  @param releaseChannel channel to pull manifests from, default is 'default'
 *  @param archivePath path to existing NSBundle to configure
 *  @param privateConfigFile path to a private config file containing, e.g., private api keys
 *  @param appleTeamId Apple Developer's account Team ID
 *  @param output specify the output path of the configured archive (ie) /tmp/my-app-archive-build.xcarchive or /tmp/my-app-ios-build.tar.gz
 *  @param type type of artifact to configure (simulator or archive)
 *  @param expoSourcePath path to expo client app sourcecode (/ios dir from expo/expo repo)
 */


let configureAndCopyArchiveAsync = (() => {
  var _ref4 = _asyncToGenerator(function* (args) {
    args = _validateCLIArgs(args);
    const { output, type } = args;
    const context = yield _createStandaloneContextAsync(args);
    yield (_IosNSBundle || _load_IosNSBundle()).configureAsync(context);
    if (output) {
      const archiveName = context.config.slug.replace(/[^0-9a-z_\-]/gi, '_');
      const appReleasePath = _path.default.resolve(context.data.archivePath, '..');
      if (type === 'simulator') {
        yield (0, (_ExponentTools || _load_ExponentTools()).spawnAsync)(`mv ${DEFAULT_EXPOKIT_WORKSPACE_NAME}.app ${archiveName}.app && tar -czvf ${output} ${archiveName}.app`, null, {
          stdoutOnly: true,
          pipeToLogger: true,
          loggerFields: { buildPhase: 'creating an archive for simulator' },
          cwd: appReleasePath,
          shell: true
        });
      } else if (type === 'archive') {
        yield (0, (_ExponentTools || _load_ExponentTools()).spawnAsync)('/bin/mv', [`${DEFAULT_EXPOKIT_WORKSPACE_NAME}.xcarchive`, output], {
          pipeToLogger: true,
          cwd: `${context.data.archivePath}/../../../..`,
          loggerFields: { buildPhase: 'renaming archive' }
        });
      }
    }
    return context.data.manifest;
  });

  return function configureAndCopyArchiveAsync(_x10) {
    return _ref4.apply(this, arguments);
  };
})();

/**
 * possible args:
 *  @param skipRepoUpdate if true, omit `--repo-update` cocoapods flag.
 */


let _createTurtleWorkspaceAsync = (() => {
  var _ref5 = _asyncToGenerator(function* (context, args) {
    const { skipRepoUpdate } = args;
    if ((_fsExtra || _load_fsExtra()).default.existsSync(context.build.ios.workspaceSourcePath)) {
      (_Logger || _load_Logger()).default.info(`Removing existing workspace at ${context.build.ios.workspaceSourcePath}...`);
      try {
        (_rimraf || _load_rimraf()).default.sync(context.build.ios.workspaceSourcePath);
      } catch (_) {}
    }
    yield (_IosWorkspace || _load_IosWorkspace()).createDetachedAsync(context);
    yield _podInstallAsync(context.build.ios.workspaceSourcePath, !skipRepoUpdate);
  });

  return function _createTurtleWorkspaceAsync(_x11, _x12) {
    return _ref5.apply(this, arguments);
  };
})();

/**
 * External-facing version can be used to create a turtle workspace without building it.
 * Probably only useful for local testing.
 *
 * @param workspacePath (optional) provide some other path to create the workspace besides the default
 * @param url (optional, with sdkVersion) url to an expo manifest, if you want the workspace to be configured automatically
 * @param sdkVersion (optional, with url) sdkVersion to an expo manifest, if you want the workspace to be configured automatically
 */


let createTurtleWorkspaceAsync = (() => {
  var _ref6 = _asyncToGenerator(function* (args) {
    args = _validateCLIArgs(args);
    if (!args.workspacePath) {
      (_Logger || _load_Logger()).default.info('No workspace path was provided with --workspacePath, so the default will be used.');
    }
    const context = yield _createStandaloneContextAsync(args);
    yield _createTurtleWorkspaceAsync(context, args);
    (_Logger || _load_Logger()).default.info(`Created turtle workspace at ${context.build.ios.workspaceSourcePath}. You can open and run this in Xcode.`);
    if (context.config) {
      yield (_IosNSBundle || _load_IosNSBundle()).configureAsync(context);
      (_Logger || _load_Logger()).default.info(`The turtle workspace was configured for the url ${args.url}. To run this app with a Debug scheme, make sure to add a development url to 'EXBuildConstants.plist'.`);
    } else {
      (_Logger || _load_Logger()).default.info(`You can specify --url <manifestUrl> --sdkVersion <version> to configure this workspace as a particular Expo app.\n\nBecause those arguments were omitted, the workspace has not been configured. It will compile but not run. The minimum configuration to get something running is to specify a manifest url in 'EXShell.plist' (for Release builds) or 'EXBuildConstants.plist' (for Debug builds).`);
    }
  });

  return function createTurtleWorkspaceAsync(_x13) {
    return _ref6.apply(this, arguments);
  };
})();

/**
 * possible args:
 *  @param configuration StandaloneBuildConfiguration (Debug or Release)
 *  @param verbose show all xcodebuild output (default false)
 *  @param reuseWorkspace if true, when building, assume a detached workspace already exists rather than creating a new one.
 *  @param type type of artifact to build (simulator or archive)
 *  @param shellAppSdkVersion sdk version for shell app
 */


let buildAndCopyArtifactAsync = (() => {
  var _ref7 = _asyncToGenerator(function* (args) {
    args = _validateCLIArgs(args);
    const context = yield _createStandaloneContextAsync(args);
    const { verbose, type, reuseWorkspace } = args;
    const { projectName } = (_IosWorkspace || _load_IosWorkspace()).getPaths(context);

    if (!reuseWorkspace) {
      yield _createTurtleWorkspaceAsync(context, args);
    }
    const pathToArtifact = yield _buildAsync(projectName, context.build.ios.workspaceSourcePath, context.build.configuration, type, _path.default.relative(context.build.ios.workspaceSourcePath, '../shellAppBase'), verbose);
    const artifactDestPath = _path.default.join('../shellAppBase-builds', type, context.build.configuration);
    (_Logger || _load_Logger()).default.info(`\nFinished building, copying artifact to ${_path.default.resolve(artifactDestPath)}...`);
    if ((_fsExtra || _load_fsExtra()).default.existsSync(artifactDestPath)) {
      yield (0, (_ExponentTools || _load_ExponentTools()).spawnAsyncThrowError)('/bin/rm', ['-rf', artifactDestPath]);
    }
    (_Logger || _load_Logger()).default.info(`mkdir -p ${artifactDestPath}`);
    yield (0, (_ExponentTools || _load_ExponentTools()).spawnAsyncThrowError)('/bin/mkdir', ['-p', artifactDestPath]);
    (_Logger || _load_Logger()).default.info(`cp -R ${pathToArtifact} ${artifactDestPath}`);
    yield (0, (_ExponentTools || _load_ExponentTools()).spawnAsyncThrowError)('/bin/cp', ['-R', pathToArtifact, artifactDestPath]);
  });

  return function buildAndCopyArtifactAsync(_x14) {
    return _ref7.apply(this, arguments);
  };
})();

var _fsExtra;

function _load_fsExtra() {
  return _fsExtra = _interopRequireDefault(require('fs-extra'));
}

var _path = _interopRequireDefault(require('path'));

var _rimraf;

function _load_rimraf() {
  return _rimraf = _interopRequireDefault(require('rimraf'));
}

var _ExponentTools;

function _load_ExponentTools() {
  return _ExponentTools = require('./ExponentTools');
}

var _IosNSBundle;

function _load_IosNSBundle() {
  return _IosNSBundle = _interopRequireWildcard(require('./IosNSBundle'));
}

var _IosWorkspace;

function _load_IosWorkspace() {
  return _IosWorkspace = _interopRequireWildcard(require('./IosWorkspace'));
}

var _StandaloneBuildFlags;

function _load_StandaloneBuildFlags() {
  return _StandaloneBuildFlags = _interopRequireDefault(require('./StandaloneBuildFlags'));
}

var _StandaloneContext;

function _load_StandaloneContext() {
  return _StandaloneContext = _interopRequireDefault(require('./StandaloneContext'));
}

var _Logger;

function _load_Logger() {
  return _Logger = _interopRequireDefault(require('./Logger'));
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

// TODO: we will need to vary this when we support multiple different build artifacts.
const DEFAULT_EXPOKIT_WORKSPACE_NAME = exports.DEFAULT_EXPOKIT_WORKSPACE_NAME = 'ExpoKitApp';

function _validateCLIArgs(args) {
  args.type = args.type || 'archive';
  args.configuration = args.configuration || 'Release';
  args.verbose = args.verbose || false;
  args.testEnvironment = args.testEnvironment || 'none';

  switch (args.type) {
    case 'simulator':
      {
        if (args.configuration !== 'Debug' && args.configuration !== 'Release') {
          throw new Error(`Unsupported build configuration ${args.configuration}`);
        }
        break;
      }
    case 'archive':
      {
        if (args.configuration !== 'Release') {
          throw new Error('Release is the only supported configuration when archiving');
        }
        break;
      }
    default:
      {
        throw new Error(`Unsupported build type ${args.type}`);
      }
  }

  switch (args.action) {
    case 'configure':
      {
        if (!args.url) {
          throw new Error('Must run with `--url MANIFEST_URL`');
        }
        if (!args.sdkVersion) {
          throw new Error('Must run with `--sdkVersion SDK_VERSION`');
        }
        if (!args.archivePath) {
          throw new Error('Need to provide --archivePath <path to existing archive for configuration>');
        }
        if (args.testEnvironment !== 'local' && args.testEnvironment !== 'ci' && args.testEnvironment !== 'none') {
          throw new Error(`Unsupported test environment ${args.testEnvironment}`);
        }
        break;
      }
    case 'build':
      {
        break;
      }
    case 'create-workspace':
      {
        break;
      }
    default:
      {
        throw new Error(`Unsupported build action ${args.action}`);
      }
  }

  return args;
}exports.buildAndCopyArtifactAsync = buildAndCopyArtifactAsync;
exports.configureAndCopyArchiveAsync = configureAndCopyArchiveAsync;
exports.createTurtleWorkspaceAsync = createTurtleWorkspaceAsync;
